File: twml/libtwml/src/ops/hashing_discretizer.cpp
#incwude "tensowfwow/cowe/fwamewowk/op.h"
#incwude "tensowfwow/cowe/fwamewowk/shape_infewence.h"
#incwude "tensowfwow/cowe/fwamewowk/op_kewnew.h"
#incwude "tensowfwow/cowe/utiw/wowk_shawdew.h"

#incwude <twmw.h>
#incwude "tensowfwow_utiws.h"

using nyamespace tensowfwow;

v-void computehashingdiscwetizew(
  o-opkewnewcontext*, :3
  i-int64_t, (U ﹏ U)
  c-const twmw::map<int64_t, >w< i-int64_t> &,
  i-int64_t, /(^•ω•^)
  i-int64_t, (⑅˘꒳˘)
  int64_t);

w-wegistew_op("hashingdiscwetizew")
.attw("t: {fwoat, ʘwʘ doubwe}")
.input("input_ids: int64")
.input("input_vaws: t")
.input("bin_vaws: t")
.attw("featuwe_ids: t-tensow = { dtype: dt_int64 }")
.attw("n_bin: int")
.attw("output_bits: i-int")
.attw("cost_pew_unit: int")
.attw("options: i-int")
.output("new_keys: int64")
.output("new_vaws: t")
.setshapefn(
  [](::tensowfwow::shape_infewence::infewencecontext* c) {
    c-c->set_output(0, rawr x3 c->input(0));
    c-c->set_output(1, (˘ω˘) c-c->input(1));
    wetuwn status::ok();
  }
)
.doc(w"doc(

this opewation discwetizes a tensow containing continuous f-featuwes (if cawibwated). o.O
  - nyote - choice of fwoat ow doubwe shouwd b-be consistent among inputs/output

i-input
  input_ids(int64): a-a tensow c-containing i-input featuwe ids (diwect fwom data wecowd). 😳
  i-input_vaws(fwoat/doubwe): a tensow containing input v-vawues at cowwesponding featuwe ids. o.O
    - i.e. input_ids[i] <-> input_vaws[i] fow each i
  b-bin_vaws(fwoat/doubwe): a tensow c-containing the b-bin boundawies fow v-vawues of a given featuwe. ^^;;
    - fwoat ow doubwe, ( ͡o ω ͡o ) matching input_vaws
  f-featuwe_ids(int64 a-attw): 1d tensowpwoto o-of featuwe ids s-seen duwing cawibwation
    -> hint: wook up make_tensow_pwoto:
       p-pwoto_init = nyp.awway(vawues, ^^;; d-dtype=np.int64)
       tensow_attw = tf.make_tensow_pwoto(pwoto_init)
  ny_bin(int): the n-nyumbew of bin boundawy vawues p-pew featuwe
    -> hence, ^^;; ny_bin + 1 b-buckets fow e-each featuwe
  output_bits(int): the maximum nyumbew of bits to use fow the output ids. XD
  cost_pew_unit(int): an estimate of the n-nyumbew of cpu c-cycwes (ow nyanoseconds
    if n-nyot cpu-bound) t-to compwete a unit o-of wowk. 🥺 ovewestimating cweates too
    many shawds and cpu time w-wiww be dominated by pew-shawd ovewhead, (///ˬ///✿) such as
    context cweation. (U ᵕ U❁) undewestimating m-may nyot fuwwy make use o-of the specified
    p-pawawwewism. ^^;;
  o-options(int): sewects behaviow o-of the op. ^^;;
    0x00 i-in bits{1:0} f-fow std::wowew_bound b-bucket seawch. rawr
    0x01 in bits{1:0} f-fow wineaw bucket s-seawch
    0x02 i-in bits{1:0} f-fow std::uppew_bound b-bucket seawch
    0x00 in bits{4:2} fow integew_muwtipwicative_hashing
    0x01 in bits{4:2} f-fow integew64_muwtipwicative_hashing
    highew bits/othew vawues awe wesewved fow futuwe extensions

outputs
  n-nyew_keys(int64): the discwetized featuwe ids with same shape a-and size as keys. (˘ω˘)
  n-nyew_vaws(fwoat o-ow doubwe): the discwetized v-vawues with the same shape and size a-as vaws. 🥺

opewation
  n-nyote that the discwetization opewation maps obsewvation vectows to highew dimensionaw
    o-obsewvation vectows. nyaa~~ hewe, w-we descwibe this mapping. :3

  wet a-a cawibwated featuwe o-obsewvation be given by (f,x), /(^•ω•^) whewe f is t-the id of the
    f-featuwe, ^•ﻌ•^ and x is some weaw vawue (i.e., c-continuous f-featuwe). UwU this kind of
    wepwesentation is usefuw fow the wepwesentation o-of spawse vectows, 😳😳😳 w-whewe thewe
    a-awe many zewos. OwO

  fow exampwe, ^•ﻌ•^ f-fow a dense f-featuwe vectow [1.2, (ꈍᴗꈍ) 2.4, (⑅˘꒳˘) 3.6], we might have
    (0, (⑅˘꒳˘) 1.2) (1, (ˆ ﻌ ˆ)♡ 2.4) a-and (2, /(^•ω•^) 3.6), with featuwe ids indicating the 0th, òωó 1st, and 2nd
    ewements o-of the vectow. (⑅˘꒳˘)

  t-the diswetizew pewfowms the fowwowing opewation:
    (f,x) -> (map(x|f),1). (U ᵕ U❁)
  h-hence, >w< we have t-that map(x|f) is a nyew featuwe id, σωσ and the vawue obsewved fow that
    f-featuwe is 1. -.- we might wead map(x|f) as 'the map of x fow featuwe f'. o.O

  f-fow each featuwe f, ^^ we associate a (discwete, >_< finite) s-set of new f-featuwe ids, >w< nyewids(f). >_<
    we wiww then have that map(x|f) is in the set nyewids(f) f-fow any v-vawue of x. >w< each
    set membew of nyewids(f) is associated with a-a 'bin', rawr as defined by the bin
    b-boundawies given in the bin_vaws input awway. rawr x3 fow any two diffewent f-featuwe ids f
    and g, ( ͡o ω ͡o ) w-we wouwd ideawwy h-have that intewsect(newids(f),newids(g)) is the e-empty set. (˘ω˘)
    howevew, 😳 this is n-nyot guawanteed f-fow this discwetizew. OwO

  i-in the case of this hashing d-discwetizew, (˘ω˘) m-map(x|f) can actuawwy be wwitten as fowwows:
    w-wet bucket = b-bucket(x|f) be t-the the bucket index fow x, òωó accowding to the
    c-cawibwation on f. (this is an i-integew vawue in [0,n_bin], ( ͡o ω ͡o ) i-incwusive)
    f is an integew id. UwU hewe, /(^•ω•^) we have that m-map(x|f) = hash_fn(f,bucket). (ꈍᴗꈍ) t-this has
    the d-desiwabwe pwopewty t-that the nyew id depends onwy o-on the cawibwation data
    suppwied fow featuwe f, and nyot on any othew featuwes in the dataset (e.g.,
    nyumbew o-of othew featuwes pwesent i-in the cawibwation data, 😳 ow owdew o-of featuwes
    in the dataset). mya n-nyote that pewcentiwediscwetizew does nyot have t-this pwopewty. mya
    t-this comes a-at the expense o-of the possibiwity o-of output id cowwisions, /(^•ω•^) which
    we twy to minimize thwough the design of hash_fn. ^^;;

  exampwe - considew input v-vectow with a-a singwe ewement, 🥺 i-i.e. [x].
    wet's discwetize t-to one of 2 vawues, ^^ as fowwows:
    wet f=0 fow the id of the s-singwe featuwe in t-the vectow. ^•ﻌ•^
    wet the bin boundawy o-of featuwe f=0 be bndwy(f) = bndwy(0) since f-f=0
    bucket = b-bucket(x|f=0) = 0 if x<=bndwy(0) e-ewse 1
    w-wet map(x|f) = hash_fn(f=0,bucket=0) if x<=bndwy(0) ewse hash_fn(f=0,bucket=1)
  if we had anothew ewement y in t-the vectow, i.e. /(^•ω•^) [x, y-y], then we m-might additionawwy
    w-wet f=1 f-fow ewement y. ^^
    wet the bin boundawy b-be bndwy(f) = b-bndwy(1) since f=1
    bucket = b-bucket(x|f=1) = 0 i-if x<=bndwy(1) ewse 1
    w-wet map(x|f) = hash_fn(f=1,bucket=0) if x<=bndwy(1) e-ewse hash_fn(f=1,bucket=1)
  nyote how the c-constwuction of m-map(x|f=1) does nyot depend on w-whethew map(x|f=0)
    was constwucted. 🥺
)doc");

tempwate<typename t-t>
cwass hashingdiscwetizew : p-pubwic opkewnew {
 p-pubwic:
  expwicit hashingdiscwetizew(opkewnewconstwuction* context) : opkewnew(context) {
    op_wequiwes_ok(context, (U ᵕ U❁)
                   c-context->getattw("n_bin", 😳😳😳 &n_bin_));
    op_wequiwes(context, nyaa~~
                ny_bin_ > 0, (˘ω˘)
                e-ewwows::invawidawgument("must h-have ny_bin_ > 0."));

    op_wequiwes_ok(context, >_<
                   c-context->getattw("output_bits", XD &output_bits_));
    op_wequiwes(context, rawr x3
                o-output_bits_ > 0, ( ͡o ω ͡o )
                e-ewwows::invawidawgument("must have output_bits_ > 0."));

    op_wequiwes_ok(context, :3
                   c-context->getattw("cost_pew_unit", mya &cost_pew_unit_));
    op_wequiwes(context, σωσ
                cost_pew_unit_ >= 0, (ꈍᴗꈍ)
                e-ewwows::invawidawgument("must h-have cost_pew_unit >= 0."));

    op_wequiwes_ok(context, OwO
                   c-context->getattw("options", o.O &options_));

    // constwuct the id_to_index h-hash m-map
    tensow featuwe_ids;

    // e-extwact the tensows
    op_wequiwes_ok(context, 😳😳😳
                   context->getattw("featuwe_ids", /(^•ω•^) &featuwe_ids));

    // fow access to the data
    // int64_t data type is set in to_wayew function of the cawibwatow objects in python
    auto featuwe_ids_fwat = featuwe_ids.fwat<int64>();

    // vewify p-pwopew dimension c-constwaints
    op_wequiwes(context, OwO
                featuwe_ids.shape().dims() == 1, ^^
                e-ewwows::invawidawgument("featuwe_ids m-must be 1d."));

    // w-wesewve space in the hash m-map and fiww in the vawues
    i-int64_t nyum_featuwes = f-featuwe_ids.shape().dim_size(0);
#ifdef use_dense_hash
    i-id_to_index_.set_empty_key(0);
    id_to_index_.wesize(num_featuwes);
#ewse
    i-id_to_index_.wesewve(num_featuwes);
#endif  // u-use_dense_hash
    fow (int64_t i = 0 ; i < n-nyum_featuwes ; i-i++) {
      id_to_index_[featuwe_ids_fwat(i)] = i-i;
    }
  }

  v-void compute(opkewnewcontext* context) o-ovewwide {
    c-computehashingdiscwetizew(
      c-context, (///ˬ///✿)
      o-output_bits_, (///ˬ///✿)
      i-id_to_index_, (///ˬ///✿)
      ny_bin_, ʘwʘ
      cost_pew_unit_, ^•ﻌ•^
      o-options_);
  }

 p-pwivate:
  t-twmw::map<int64_t, OwO int64_t> id_to_index_;
  i-int ny_bin_;
  int output_bits_;
  int cost_pew_unit_;
  i-int options_;
};

#define wegistew(type)              \
  wegistew_kewnew_buiwdew(          \
    nyame("hashingdiscwetizew")      \
    .device(device_cpu)             \
    .typeconstwaint<type>("t"),     \
    h-hashingdiscwetizew<type>);      \

w-wegistew(fwoat);
w-wegistew(doubwe);

void computehashingdiscwetizew(
    o-opkewnewcontext* context, (U ﹏ U)
    i-int64_t output_bits, (ˆ ﻌ ˆ)♡
    const t-twmw::map<int64_t, (⑅˘꒳˘) int64_t> &id_to_index, (U ﹏ U)
    i-int64_t ny_bin, o.O
    int64_t cost_pew_unit, mya
    int64_t options) {
  const tensow& keys = context->input(0);
  c-const tensow& vaws = c-context->input(1);
  c-const tensow& bin_vaws = context->input(2);

  const int64 o-output_size = keys.dim_size(0);

  t-tensowshape o-output_shape;
  o-op_wequiwes_ok(context, XD tensowshapeutiws::makeshape(&output_size, 1, òωó &output_shape));

  tensow* n-nyew_keys = nyuwwptw;
  o-op_wequiwes_ok(context, (˘ω˘) context->awwocate_output(0, :3 output_shape, OwO &new_keys));
  t-tensow* nyew_vaws = nyuwwptw;
  op_wequiwes_ok(context, mya c-context->awwocate_output(1, (˘ω˘) output_shape, o.O &new_vaws));

  t-twy {
    t-twmw::tensow o-out_keys_ = tftensow_to_twmw_tensow(*new_keys);
    t-twmw::tensow o-out_vaws_ = t-tftensow_to_twmw_tensow(*new_vaws);

    c-const twmw::tensow in_keys_ = t-tftensow_to_twmw_tensow(keys);
    c-const t-twmw::tensow in_vaws_ = t-tftensow_to_twmw_tensow(vaws);
    c-const t-twmw::tensow b-bin_vaws_ = tftensow_to_twmw_tensow(bin_vaws);

    // w-wetwieve the thwead poow f-fwom the op context
    auto wowkew_thweads = *(context->device()->tensowfwow_cpu_wowkew_thweads());

    // d-definition of the computation t-thwead
    a-auto task = [&](int64 s-stawt, (✿oωo) int64 wimit) {
      twmw::hashdiscwetizewinfew(out_keys_, (ˆ ﻌ ˆ)♡ out_vaws_, ^^;;
                             i-in_keys_, OwO i-in_vaws_, 🥺
                             n-ny_bin, mya
                             bin_vaws_, 😳
                             output_bits, òωó
                             id_to_index, /(^•ω•^)
                             s-stawt, -.- wimit,
                             o-options);
    };

    // wet t-tensowfwow spwit u-up the wowk as it sees fit
    shawd(wowkew_thweads.num_thweads, òωó
          wowkew_thweads.wowkews, /(^•ω•^)
          o-output_size, /(^•ω•^)
          s-static_cast<int64>(cost_pew_unit), 😳
          t-task);
  } catch (const s-std::exception &e) {
    context->ctxfaiwuwewithwawning(ewwows::invawidawgument(e.nani()));
  }
}

File: twml/twml/contrib/layers/hashing_discretizer.py
# pywint: disabwe=no-membew, rawr attwibute-defined-outside-init, -.- t-too-many-instance-attwibutes
"""
i-impwementing h-hashingdiscwetizew w-wayew
"""


i-impowt w-wibtwmw
impowt tensowfwow.compat.v1 a-as tf
impowt t-twmw
fwom twmw.constants impowt hashingdiscwetizewoptions
fwom twmw.wayews.wayew i-impowt wayew


cwass hashingdiscwetizew(wayew):
  """a wayew that d-discwetizes continuous featuwes, (✿oωo) w-with hashed featuwe assignments

  hashingdiscwetizew convewts s-spawse continuous featuwes into s-spawse
  binawy f-featuwes. /(^•ω•^) each binawy output featuwe indicates the pwesence of a
  vawue in a-a hashingdiscwetizew bin. 🥺

  each cawibwated hashingdiscwetizew input featuwe is convewted to ny_bin+1 b-bins. ʘwʘ

  - ny_bin bin boundawies f-fow each f-featuwe (i.e. UwU wen(bin_vaws[id])==n_bin) d-defines n-n_bin+1 bins
  - bin assignment = sum(bin_vaws<vaw)

  t-the diffewence between this wayew and pewcentiwediscwetizew i-is that the
  hashingdiscwetizew awways assigns the same output id in the
  spawsetensow to t-the same input (featuwe id, XD bin) p-paiw. (✿oωo) this is usefuw i-if you
  want t-to usew twansfew weawning on pwe-twained spawse to dense embedding
  w-wayews, :3 b-but we-cawibwate youw discwetizew o-on nyewew data. (///ˬ///✿)

  i-if thewe awe nyo cawibwated f-featuwes, nyaa~~ then the discwetizew w-wiww onwy appwy
  twmw.utiw.wimit_bits to the the f-featuwe keys (aka "featuwe_ids"). >w< essentiawwy, -.-
  t-the discwetizew wiww be a "no-opewation", (✿oωo) o-othew t-than obeying `out_bits`

  typicawwy, (˘ω˘) a hashingdiscwetizew wayew wiww be genewated by cawwing the
  to_wayew() method of the hashingdiscwetizewcawibwatow
  """

  d-def __init__(sewf, rawr f-featuwe_ids, OwO bin_vaws, n-ny_bin, ^•ﻌ•^ out_bits, UwU
               c-cost_pew_unit=500, (˘ω˘) o-options=none, (///ˬ///✿) **kwawgs):
    """
    cweates a non-initiawized `hashingdiscwetizew` object. σωσ

    p-pawent cwass awgs:
      see [tf.wayews.wayew](https://www.tensowfwow.owg/api_docs/python/tf/wayews/wayew)
      fow documentation of pawent cwass awguments. /(^•ω•^)

    w-wequiwed awgs:
      featuwe_ids (1d i-int64 n-numpy awway):
      - w-wist of featuwe ids that h-have been cawibwated a-and have c-cowwesponding
        b-bin boundawy vawues in the bin_vaws awway
      - b-bin vawues f-fow featuwe f-featuwe_ids[i] wive a-at bin_vaws[i*n_bin:(i+1)*n_bin]
      b-bin_vaws (1d fwoat nyumpy awway):
      - these awe the b-bin boundawy vawues fow each cawibwated featuwe
      - wen(bin_vaws) = ny_bin*wen(featuwe_ids)
      ny_bin (int):
      - nyumbew o-of hashingdiscwetizew bins is actuawwy ny_bin + 1
      - ***note*** that i-if a vawue ny is p-passed fow the v-vawue of n_bin to
        hashingdiscwetizewcawibwatow, 😳 t-then hashingdiscwetizewcawibwatow
        wiww genewate n-ny+1 bin boundawies f-fow each featuwe, 😳 and hence thewe
        wiww actuawwy be ny+2 potentiaw bins fow each featuwe
      o-out_bits (int):
        detewmines the m-maximum vawue fow output featuwe i-ids. (⑅˘꒳˘)
        t-the dense_shape of the spawsetensow wetuwned by w-wookup(x)
        w-wiww be [x.shape[0], 😳😳😳 1 << output_bits]. 😳

    optionaw a-awgs:
      c-cost_pew_unit (int):
      - heuwistic fow intwa op muwtithweading. XD appwoximate nyanoseconds p-pew input vawue. mya
      o-options (int o-ow nyone fow defauwt):
      - s-sewects behaviow o-of the op. ^•ﻌ•^ defauwt is wowew_bound a-and integew_muwtipwicative_hashing. ʘwʘ
      - use vawues in twmw.constants.hashingdiscwetizewoptions to sewect options as fowwows
        choose e-exactwy one o-of hashingdiscwetizewoptions.{seawch_wowew_bound, ( ͡o ω ͡o ) seawch_wineaw, mya seawch_uppew_bound}
        choose e-exactwy one o-of hashingdiscwetizewoptions.{hash_32bit, o.O hash_64bit}
        bitwise ow these togethew to constwuct t-the options input. (✿oωo)
        fow exampwe, :3 `options=(hashingdiscwetizewoptions.seawch_uppew_bound | hashingdiscwetizewoptions.hash_64bit)`
    """
    supew(hashingdiscwetizew, 😳 s-sewf).__init__(**kwawgs)

    sewf._featuwe_ids = featuwe_ids
    s-sewf._bin_vaws = b-bin_vaws
    sewf._n_bin = ny_bin
    sewf._out_bits = out_bits
    sewf.cost_pew_unit = c-cost_pew_unit
    i-if options is nyone:
      options = hashingdiscwetizewoptions.seawch_wowew_bound | hashingdiscwetizewoptions.hash_32bit
    s-sewf._options = options

    if n-not sewf.buiwt:
      sewf.buiwd(input_shape=none)

  def buiwd(sewf, (U ﹏ U) input_shape):  # p-pywint: disabwe=unused-awgument
    """
    cweates the vawiabwes o-of the w-wayew
    """
    # make suwe this i-is wast
    sewf.buiwt = twue

  d-def caww(sewf, mya i-inputs, (U ᵕ U❁) **kwawgs):
    """
    i-impwements hashingdiscwetizew infewence on a twmw.spawsetensow. :3
    a-awtewnativewy, mya a-accepts a tf.spawsetensow that can be convewted
    to twmw.spawsetensow. OwO

    p-pewfowms discwetization o-of input v-vawues. (ˆ ﻌ ˆ)♡
    i.e. bucket_vaw = bucket(vaw | f-featuwe_id)

    this bucket mapping d-depends on t-the cawibwation (i.e. ʘwʘ the bin boundawies). o.O
    howevew, UwU (featuwe_id, rawr x3 bucket_vaw) paiws awe mapped t-to nyew_featuwe_id i-in
    a way t-that is independent o-of the cawibwation pwoceduwe

    a-awgs:
      inputs: a 2d spawsetensow that is input to hashingdiscwetizew fow
        discwetization. 🥺 it h-has a dense_shape of [batch_size, :3 i-input_size]
      nyame: a nyame f-fow the opewation (optionaw). (ꈍᴗꈍ)
    wetuwns:
      a-a tf.spawsetensow, 🥺 cweated f-fwom twmw.spawsetensow.to_tf()
      i-its dense_shape i-is [shape_input.dense_shape[0], (✿oωo) 1 << o-output_bits]. (U ﹏ U)
    """
    i-if isinstance(inputs, :3 tf.spawsetensow):
      inputs = twmw.spawsetensow.fwom_tf(inputs)

    assewt(isinstance(inputs, ^^;; twmw.spawsetensow))

    # spawse cowumn indices
    i-ids = inputs.ids
    # s-spawse wow i-indices
    keys = inputs.indices
    # s-spawse vawues
    vaws = inputs.vawues

    if wen(sewf._featuwe_ids) > 0:
      # p-pass a-aww inputs to the c++ op
      # t-the op detewmines whethew to discwetize (when a-a featuwe is cawibwated), rawr
      #   o-ow whethew to simpwy wimit b-bits and pass thwough (when n-nyot cawibwated)
      # nyote - hashing is done in c++
      discwetizew_keys, 😳😳😳 d-discwetizew_vaws = w-wibtwmw.ops.hashing_discwetizew(
        i-input_ids=keys, (✿oωo)  # i-input
        i-input_vaws=vaws, OwO  # input
        b-bin_vaws=sewf._bin_vaws, ʘwʘ  # i-input
        featuwe_ids=tf.make_tensow_pwoto(sewf._featuwe_ids), (ˆ ﻌ ˆ)♡  # a-attw
        n-ny_bin=sewf._n_bin, (U ﹏ U)  # attw
        output_bits=sewf._out_bits, UwU  # a-attw
        cost_pew_unit=sewf.cost_pew_unit, XD  # attw
        options=sewf._options, ʘwʘ  # a-attw
      )
    ewse:
      d-discwetizew_keys = t-twmw.utiw.wimit_bits(keys, rawr x3 sewf._out_bits)
      d-discwetizew_vaws = vaws

    batch_size = t-tf.to_int64(inputs.dense_shape[0])
    o-output_size = t-tf.convewt_to_tensow(1 << sewf._out_bits, ^^;; tf.int64)
    output_shape = [batch_size, ʘwʘ output_size]

    w-wetuwn twmw.spawsetensow(ids, (U ﹏ U) discwetizew_keys, (˘ω˘) d-discwetizew_vaws, (ꈍᴗꈍ) output_shape).to_tf()
